libs/collectives/include/hpx/collectives/all_reduce.hpp

```diff
@@ -0,0 +1,512 @@
+    ///
+    /// The macro \a HPX_REGISTER_ALLREDUCE can be used to define
+    /// all facilities necessary for a (possibly remote) all_reduce operation.
+    ///
+    /// The parameter \a type specifies for which data type the all_reduce
+    /// operations should be enabled.
+    ///
+    /// The (optional) parameter \a name should be a unique C-style identifier
+    /// that will be internally used to identify a particular all_reduce operation.
+    /// If this defaults to \a \<type\>_all_reduce if not specified.
+    ///
+    /// \note The macro \a HPX_REGISTER_ALLREDUCE can be used with 1
+    ///       or 2 arguments. The second argument is optional and defaults to
+    ///       \a \<type\>_all_reduce.
+    ///
+    #define HPX_REGISTER_ALLREDUCE(type, name)
+}}
+#else
+
+#include <hpx/config.hpp>
+
+#if !defined(HPX_COMPUTE_DEVICE_CODE)
+
+#include <hpx/dataflow.hpp>
+#include <hpx/lcos/future.hpp>
+#include <hpx/lcos/local/and_gate.hpp>
+#include <hpx/lcos/local/spinlock.hpp>
+#include <hpx/parallel/algorithms/reduce.hpp>
+#include <hpx/preprocessor/cat.hpp>
+#include <hpx/preprocessor/expand.hpp>
+#include <hpx/preprocessor/nargs.hpp>
+#include <hpx/runtime/basename_registration.hpp>
+#include <hpx/runtime/components/new.hpp>
+#include <hpx/runtime/components/server/component_base.hpp>
+#include <hpx/runtime/get_num_localities.hpp>
+#include <hpx/runtime/launch_policy.hpp>
+#include <hpx/runtime/naming/id_type.hpp>
+#include <hpx/runtime/naming/unmanaged.hpp>
+#include <hpx/util/assert.hpp>
+#include <hpx/util/decay.hpp>
+#include <hpx/util/unused.hpp>
+
+#include <cstddef>
+#include <mutex>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace hpx { namespace lcos
+{
+    namespace detail
+    {
+        ////////////////////////////////////////////////////////////////////////
+        template <typename T>
+        class all_reduce_server
+          : public hpx::components::component_base<all_reduce_server<T> >
+        {
+            using mutex_type = lcos::local::spinlock;
+
+            template <typename F>
+            T on_ready(shared_future<void>&& f)
+            {
+                f.get();       // propagate any exceptions
+
+                std::vector<T> data;
+                std::string name;
+
+                {
+                    std::unique_lock<mutex_type> l(mtx_);
+                    data = data_;
+                    std::swap(name, name_);
+                }
+
+                // this is a one-shot object (generations counters are not
+                // supported), unregister ourselves (but only once)
+                if (!name.empty())
+                {
+                    hpx::unregister_with_basename(name, site_);
+                }
+
+                return hpx::parallel::reduce(hpx::parallel::execution::par,
+                    data.begin(), data.end(), std::forward<F>(op));
```

Comment:
Where is `op` defined? https://circleci.com/gh/STEllAR-GROUP/hpx/102137?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link

Reply:
This is an oversight, the function is not needed anymore. Will fix.